apps/anoma_node/lib/examples/e_shard_backend.ex

```diff
@@ -0,0 +1,718 @@
+    :ok = wait_for_block(node_id, 1)
+    Logger.info("Finished executing Tx1.")
+    log_shard_states(node_id, "After Tx1")
+
+    # Wait for the write watermark on shard 'a' to be updated by Ordering
+    wait_for_state_condition(
+      Registry.whereis(node_id, Shard, :a),
+      "write WM on a == 1 after Tx1",
+      fn state ->
+        get_in(state.watermarks, ["a", :write]) == 1
+      end
+    )
+
+    # Verify watermarks after Tx1
+    state_a_after_tx1 = :sys.get_state(Registry.whereis(node_id, Shard, :a))
+
+    assert state_a_after_tx1.watermarks["a"].write == 1,
+           "Shard 'a' write watermark should be 1 after Tx1"
+
+    # Execute Tx2
+    Logger.info("Executing Tx2...")
+    :ok = Mempool.execute(node_id, ["tx2"])
+    :ok = wait_for_block(node_id, 2)
+    Logger.info("Finished executing Tx2.")
+    log_shard_states(node_id, "After Tx2")
+
+    # Wait for the write watermark on shard 'b' to be updated by Ordering
+    wait_for_state_condition(
+      Registry.whereis(node_id, Shard, :b),
+      "write WM on b == 2 after Tx2",
+      fn state ->
+        get_in(state.watermarks, ["b", :write]) == 2
+      end
+    )
+
+    # Verify watermarks after Tx2
+    state_b_after_tx2 = :sys.get_state(Registry.whereis(node_id, Shard, :b))
+
+    assert state_b_after_tx2.watermarks["b"].write == 2,
+           "Shard 'b' write watermark should be 2 after Tx2"
+
+    # Execute Tx3
+    Logger.info("Executing Tx3...")
+    :ok = Mempool.execute(node_id, ["tx3"])
+    :ok = wait_for_block(node_id, 3)
+    Logger.info("Finished executing Tx3.")
+    log_shard_states(node_id, "After Tx3 (Final)")
+
+    # Wait for the final watermarks to be updated by Ordering
+    wait_for_state_condition(
+      Registry.whereis(node_id, Shard, :a),
+      "read WM on a == 3 after Tx3",
+      fn state ->
+        get_in(state.watermarks, ["a", :read]) == 3
+      end
+    )
+
+    wait_for_state_condition(
+      Registry.whereis(node_id, Shard, :b),
+      "read WM on b == 3 after Tx3",
+      fn state ->
+        get_in(state.watermarks, ["b", :read]) == 3
+      end
+    )
+
+    wait_for_state_condition(
+      Registry.whereis(node_id, Shard, :c),
+      "write WM on c == 3 after Tx3",
+      fn state ->
+        get_in(state.watermarks, ["c", :write]) == 3
+      end
+    )
+
+    # 5. Verification
+    # Get shard PIDs using the atom labels used in schema/supervisor
+    pid_a = Registry.whereis(node_id, Shard, :a)
+    pid_b = Registry.whereis(node_id, Shard, :b)
+    pid_c = Registry.whereis(node_id, Shard, :c)
+
+    assert is_pid(pid_a),
+           "Shard 'a' PID not found. ShardSupervisor might not have started it correctly."
+
+    assert is_pid(pid_b), "Shard 'b' PID not found."
+    assert is_pid(pid_c), "Shard 'c' PID not found."
+
+    # Check final state using :sys.get_state - Tx1 (h=1), Tx2 (h=2), Tx3 (h=3)
+    state_a = :sys.get_state(pid_a)
+    state_b = :sys.get_state(pid_b)
+    state_c = :sys.get_state(pid_c)
```

Comment:
We are abusing `get_state` we should not use it this much unless it's for meta operations, and I don't believe this counts, can we check this another way?